funsor/distributions.py

```diff
@@ -88,68 +88,98 @@ def eager_reduce(self, op, reduced_vars):
         return super(Distribution, self).eager_reduce(op, reduced_vars)
 
     @classmethod
-    def eager_log_prob(cls, **params):
-        inputs, tensors = align_tensors(*params.values())
-        params = dict(zip(params, tensors))
+    def eager_log_prob(cls, *params):
+        inputs, tensors = align_tensors(*params)
+        params = dict(zip(cls._ast_fields, tensors))
         value = params.pop('value')
         data = cls.dist_class(**params).log_prob(value)
         return Tensor(data, inputs)
 
+    def __getattribute__(self, attr):
+        if attr in type(self)._ast_fields and attr != 'name':
+            return self.params[attr]
+        return super().__getattribute__(attr)
+
+    @classmethod
+    def _infer_value_domain(cls, **kwargs):
+        # rely on the underlying distribution's logic to infer the event_shape
+        instance = cls.dist_class(**{k: _dummy_tensor(v.output) for k, v in kwargs.items()}, validate_args=False)
+        out_shape = instance.event_shape
+        if isinstance(instance.support, torch.distributions.constraints._IntegerInterval):
+            out_dtype = int(instance.support.upper_bound + 1)
+            # this is a hack, but we don't really care about precise dtypes except for Categorical
+            out_dtype = 'real' if out_dtype == 1 else out_dtype
+        else:
+            out_dtype = 'real'
+        return Domain(dtype=out_dtype, shape=out_shape)
+
 
 ################################################################################
 # Distribution Wrappers
 ################################################################################
 
-class BernoulliProbs(Distribution):
-    """
-    Wraps :class:`pyro.distributions.Bernoulli` .
+def make_dist(pyro_dist_class, param_names=()):
 
-    :param Funsor probs: Probability of 1.
-    :param Funsor value: Optional observation in ``{0,1}``.
-    """
-    dist_class = dist.Bernoulli
+    if not param_names:
+        param_names = tuple(pyro_dist_class.arg_constraints.keys())
 
-    @staticmethod
-    def _fill_defaults(probs, value='value'):
-        probs = to_funsor(probs)
-        assert probs.dtype == "real"
-        value = to_funsor(value, reals())
-        return probs, value
+    @makefun.with_signature(f"__init__(self, {', '.join(param_names)}, value='value')")
+    def dist_init(self, *args, **kwargs):
+        return Distribution.__init__(self, *tuple(kwargs.values()))
```

Comment:
Isn't this nondeterministic? Why not simply pass through `**kwargs`?

Reply:
This should be safe in Python 3.6+ but might as well pass `**kwargs` as it's clearer.